File: src/decider.js
import { spawn } from 'node:child_process';

// Ask an external command (a script, or an LLM CLI like `claude -p`) what to do
// about a suspected prompt. The context is written to the command's stdin; the
// command must print exactly one directive on its first stdout line:
//
//   SEND:<text>   type <text> and press Enter
//   ENTER         just press Enter
//   CANCEL        abort the child process (Ctrl+C)
//   WAIT          not a prompt / still working — do nothing
//
// Anything else is treated as WAIT.

const INSTRUCTIONS = `You are supervising a terminal program that appears to be waiting for input.
Below is the tail of its output. Decide how to respond.
Reply with EXACTLY ONE line and nothing else, in one of these forms:
SEND:<text>   (type <text> and press Enter — e.g. SEND:y or SEND:my-project)
ENTER         (just press Enter)
CANCEL        (abort the program; use for password prompts or anything unsafe)
WAIT          (it is not actually waiting for input)

--- terminal output tail ---
`;

// Credential-choice mode: the decider is shown the available credential *names*
// (never the secret values) and picks which one fits the prompt.
const CRED_INSTRUCTIONS = (refs) => `You are supervising a terminal program that is asking for a credential (a password, passphrase, or token).
Below is the tail of its output. Choose which stored credential should be used.
Available credentials — NAMES ONLY, you never see their values: ${refs.join(', ')}
Reply with EXACTLY ONE line and nothing else:
CRED:<name>   (use this credential — <name> MUST be one of the list above)
CANCEL        (do not provide any credential; use if none clearly fits or it's unsafe)

--- terminal output tail ---
`;

function runDecider(deciderCmd, input, timeoutMs, parse) {
  return new Promise((resolve) => {
    const child = spawn(deciderCmd, {
      shell: true,
      stdio: ['pipe', 'pipe', 'inherit'],
      windowsHide: true,
    });

    let out = '';
    let done = false;
    const finish = (verdict) => {
      if (done) return;
      done = true;
      clearTimeout(timer);
      try { child.kill(); } catch {}
      resolve(verdict);
    };

    const timer = setTimeout(() => finish({ type: 'WAIT', raw: '(decider timeout)' }), timeoutMs);

    child.stdout.on('data', (d) => { out += d.toString(); });
    child.on('error', (err) => finish({ type: 'WAIT', raw: `(decider error: ${err.message})` }));
    child.on('close', () => finish(parse(out)));

    child.stdin.write(input);
    child.stdin.end();
  });
}

export function askDecider(deciderCmd, tail, timeoutMs = 60_000) {
  return runDecider(deciderCmd, INSTRUCTIONS + tail + '\n', timeoutMs, parseVerdict);
}

export function askCredentialChoice(deciderCmd, tail, refs, timeoutMs = 60_000) {
  return runDecider(deciderCmd, CRED_INSTRUCTIONS(refs) + tail + '\n', timeoutMs, parseCredVerdict);
}

export function parseCredVerdict(output) {
  const line = output
    .split('\n')
    .map((l) => l.trim())
    .find((l) => /^(CRED:|CANCEL$)/.test(l));
  if (!line) return { type: 'WAIT', raw: output.trim().slice(0, 200) };
  if (line.startsWith('CRED:')) return { type: 'CRED', ref: line.slice(5).trim(), raw: line };
  return { type: 'CANCEL', raw: line };
}

export function parseVerdict(output) {
  const line = output
    .split('\n')
    .map((l) => l.trim())
    .find((l) => /^(SEND:|ENTER$|CANCEL$|WAIT$)/.test(l));
  if (!line) return { type: 'WAIT', raw: output.trim().slice(0, 200) };
  if (line.startsWith('SEND:')) return { type: 'SEND', text: line.slice(5), raw: line };
  return { type: line, raw: line };
}

File: engine-rs/src/decider.rs
use std::process::Stdio;
use std::time::Duration;

use tokio::io::AsyncWriteExt;

// Ask an external command (a script, or an LLM CLI like `claude -p`) what to
// do about a suspected prompt. The context goes to the command's stdin; the
// command prints exactly one directive line:
//   SEND:<text> / ENTER / CANCEL / WAIT     (prompt mode)
//   CRED:<name> / CANCEL                    (credential-choice mode)
// Anything else is treated as WAIT.

const INSTRUCTIONS: &str =
    "You are supervising a terminal program that appears to be waiting for input.
Below is the tail of its output. Decide how to respond.
Reply with EXACTLY ONE line and nothing else, in one of these forms:
SEND:<text>   (type <text> and press Enter — e.g. SEND:y or SEND:my-project)
ENTER         (just press Enter)
CANCEL        (abort the program; use for password prompts or anything unsafe)
WAIT          (it is not actually waiting for input)

--- terminal output tail ---
";

fn cred_instructions(refs: &[String]) -> String {
    format!(
        "You are supervising a terminal program that is asking for a credential (a password, passphrase, or token).
Below is the tail of its output. Choose which stored credential should be used.
Available credentials — NAMES ONLY, you never see their values: {}
Reply with EXACTLY ONE line and nothing else:
CRED:<name>   (use this credential — <name> MUST be one of the list above)
CANCEL        (do not provide any credential; use if none clearly fits or it's unsafe)

--- terminal output tail ---
",
        refs.join(", ")
    )
}

pub struct Verdict {
    pub kind: VerdictKind,
    pub raw: String,
}

pub enum VerdictKind {
    Send(String),
    Enter,
    Cancel,
    Wait,
    Cred(String),
}

async fn run_decider(decider_cmd: &str, input: String, timeout: Duration) -> String {
    let run = async {
        #[cfg(windows)]
        let mut cmd = {
            let mut c = tokio::process::Command::new("cmd");
            c.arg("/C").arg(decider_cmd);
            c.creation_flags(0x0800_0000); // CREATE_NO_WINDOW
            c
        };
        #[cfg(not(windows))]
        let mut cmd = {
            let mut c = tokio::process::Command::new("sh");
            c.arg("-c").arg(decider_cmd);
            c
        };
        cmd.stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::inherit());
        let mut child = cmd.spawn().map_err(|e| format!("(decider error: {e})"))?;
        if let Some(mut stdin) = child.stdin.take() {
            let _ = stdin.write_all(input.as_bytes()).await;
            drop(stdin);
        }
        let out = child
            .wait_with_output()
            .await
            .map_err(|e| format!("(decider error: {e})"))?;
        Ok::<String, String>(String::from_utf8_lossy(&out.stdout).into_owned())
    };
    match tokio::time::timeout(timeout, run).await {
        Ok(Ok(out)) => out,
        Ok(Err(e)) => e,
        Err(_) => "(decider timeout)".into(),
    }
}

pub async fn ask_decider(decider_cmd: &str, tail: &str) -> Verdict {
    let out = run_decider(
        decider_cmd,
        format!("{INSTRUCTIONS}{tail}\n"),
        Duration::from_secs(60),
    )
    .await;
    parse_verdict(&out)
}

pub async fn ask_credential_choice(decider_cmd: &str, tail: &str, refs: &[String]) -> Verdict {
    let out = run_decider(
        decider_cmd,
        format!("{}{tail}\n", cred_instructions(refs)),
        Duration::from_secs(60),
    )
    .await;
    parse_cred_verdict(&out)
}

pub fn parse_verdict(output: &str) -> Verdict {
    let line = output
        .lines()
        .map(str::trim)
        .find(|l| l.starts_with("SEND:") || *l == "ENTER" || *l == "CANCEL" || *l == "WAIT");
    match line {
        Some(l) if l.starts_with("SEND:") => Verdict {
            kind: VerdictKind::Send(l[5..].to_string()),
            raw: l.to_string(),
        },
        Some("ENTER") => Verdict {
            kind: VerdictKind::Enter,
            raw: "ENTER".into(),
        },
        Some("CANCEL") => Verdict {
            kind: VerdictKind::Cancel,
            raw: "CANCEL".into(),
        },
        Some(l) => Verdict {
            kind: VerdictKind::Wait,
            raw: l.to_string(),
        },
        None => Verdict {
            kind: VerdictKind::Wait,
            raw: output.trim().chars().take(200).collect(),
        },
    }
}

pub fn parse_cred_verdict(output: &str) -> Verdict {
    let line = output
        .lines()
        .map(str::trim)
        .find(|l| l.starts_with("CRED:") || *l == "CANCEL");
    match line {
        Some(l) if l.starts_with("CRED:") => Verdict {
            kind: VerdictKind::Cred(l[5..].trim().to_string()),
            raw: l.to_string(),
        },
        Some(_) => Verdict {
            kind: VerdictKind::Cancel,
            raw: "CANCEL".into(),
        },
        None => Verdict {
            kind: VerdictKind::Wait,
            raw: output.trim().chars().take(200).collect(),
        },
    }
}
